File: src/main/scala/com/exasol/cloudetl/kafka/deserialization/FieldParser.scala
package com.exasol.cloudetl.kafka.deserialization

import com.exasol.cloudetl.kafka.KafkaConnectorException
import com.exasol.errorreporting.ExaError

/**
 * Parse the fields specified to be extracted from a single Kafka record.
 */
object FieldParser {

  private val FieldReference = raw"(key|value)\.([\S]+)+".r

  def get(fields: Seq[String]): Seq[GlobalFieldSpecification] =
    fields.map {
      case "key.*"     => RecordKeyFields
      case "value.*"   => RecordValueFields
      case "timestamp" => TimestampField
      case "key"       => RecordKey
      case "value"     => RecordValue
      case anythingElse =>
        anythingElse match {
          case FieldReference(keyOrValue, fieldName) =>
            keyOrValue match {
              case "key"   => RecordKeyField(fieldName)
              case "value" => RecordValueField(fieldName)
              case _ =>
                throw new KafkaConnectorException(
                  ExaError
                    .messageBuilder("E-KCE-13")
                    .message("Field reference can only contain 'key' or 'value' fields.")
                    .mitigation("Please check that the provided field reference is correct.")
                    .toString()
                )
            }
          case _ =>
            throw new KafkaConnectorException(
              ExaError
                .messageBuilder("E-KCE-14")
                .message(s"Field reference {{REFERENCE}} does not have the correct format.", anythingElse)
                .mitigation(
                  "It must be one of [key, value, key.*, value.*, key.fieldName, value.fieldName, timestamp] values."
                )
                .toString()
            )
        }
    }

  def get(columnString: String): Seq[GlobalFieldSpecification] =
    get(columnString.split(",").map(_.trim).toSeq)
}

File: src/main/java/com/exasol/cloudetl/kafka/deserialization/FieldParser.java
package com.exasol.cloudetl.kafka.deserialization;

import java.util.ArrayList;
import java.util.List;

import com.exasol.cloudetl.kafka.KafkaConnectorException;
import com.exasol.cloudetl.kafka.ScalaCollections;
import com.exasol.errorreporting.ExaError;

public final class FieldParser {
    private FieldParser() {
    }

    public static List<GlobalFieldSpecification> get(final List<String> fields) {
        final List<GlobalFieldSpecification> result = new ArrayList<>();
        for (final String field : fields) {
            result.add(parse(field));
        }
        return result;
    }

    public static List<GlobalFieldSpecification> get(final scala.collection.immutable.Seq<String> fields) {
        return get(ScalaCollections.javaList(fields));
    }

    public static List<GlobalFieldSpecification> get(final String columnString) {
        final List<String> fields = new ArrayList<>();
        for (final String field : columnString.split(",")) {
            fields.add(field.trim());
        }
        return get(fields);
    }

    private static GlobalFieldSpecification parse(final String field) {
        switch (field) {
            case "key.*":
                return FieldSpecificationSingletons.recordKeyFields();
            case "value.*":
                return FieldSpecificationSingletons.recordValueFields();
            case "timestamp":
                return FieldSpecificationSingletons.timestampField();
            case "key":
                return FieldSpecificationSingletons.recordKey();
            case "value":
                return FieldSpecificationSingletons.recordValue();
            default:
                return parseConcreteField(field);
        }
    }

    private static GlobalFieldSpecification parseConcreteField(final String field) {
        final int separator = field.indexOf('.');
        if (separator <= 0 || separator == field.length() - 1 || hasWhitespace(field)) {
            throw invalidFormat(field);
        }
        final String recordPart = field.substring(0, separator);
        final String fieldName = field.substring(separator + 1);
        if ("key".equals(recordPart)) {
            return new RecordKeyField(fieldName);
        } else if ("value".equals(recordPart)) {
            return new RecordValueField(fieldName);
        } else {
            throw invalidFormat(field);
        }
    }

    private static boolean hasWhitespace(final String value) {
        for (int index = 0; index < value.length(); index++) {
            if (Character.isWhitespace(value.charAt(index))) {
                return true;
            }
        }
        return false;
    }

    private static KafkaConnectorException invalidFormat(final String field) {
        return new KafkaConnectorException(ExaError.messageBuilder("E-KCE-14")
                .message("Field reference {{REFERENCE}} does not have the correct format.", field)
                .mitigation("It must be one of [key, value, key.*, value.*, key.fieldName, value.fieldName, timestamp] values.")
                .toString());
    }
}
